packages/firestore/test/integration/remote/stream.test.ts

```diff
@@ -0,0 +1,231 @@
+/**
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { expect } from 'chai';
+import { EmptyCredentialsProvider } from '../../../src/api/credentials';
+import { SnapshotVersion } from '../../../src/core/snapshot_version';
+import { MutationResult } from '../../../src/model/mutation';
+import { PlatformSupport } from '../../../src/platform/platform';
+import { Datastore } from '../../../src/remote/datastore';
+import {
+  PersistentListenStream,
+  PersistentWriteStream,
+  WatchStreamListener,
+  WriteStreamListener
+} from '../../../src/remote/persistent_stream';
+import {
+  DocumentWatchChange,
+  ExistenceFilterChange,
+  WatchTargetChange
+} from '../../../src/remote/watch_change';
+import { AsyncQueue } from '../../../src/util/async_queue';
+import { Deferred } from '../../../src/util/promise';
+import { asyncIt } from '../../util/helpers';
+import { getDefaultDatabaseInfo } from '../util/helpers';
+import FirestoreError = firestore.FirestoreError;
+
+type StreamStatusCallback =
+  | 'onHandshakeComplete'
+  | 'onMutationResult'
+  | 'onWatchChange'
+  | 'onOpen'
+  | 'onClose';
+
+class StreamStatusListener implements WatchStreamListener, WriteStreamListener {
+  private pendingStates: StreamStatusCallback[] = [];
+  private pendingPromises: Deferred<StreamStatusCallback>[] = [];
+
+  private resolvePending(actualCallback: StreamStatusCallback) {
+    let pendingPromise = this.pendingPromises.shift();
```

Comment:
nit: Here you're calling shift() and checking for undefined.  Below (for pendingStates) you check the length > 0 before calling shift.  I prefer the latter, but regardless I'd try to be consistent.

Reply:
I actually prefer the undefined check since it natural leads to the declaration of another variable, which makes the code a little clearer in my opinion.